internal/clientconn/cursor/cursor.go

```diff
@@ -67,30 +98,81 @@ func newCursor(id int64, iter types.DocumentsIterator, params *NewParams, r *Reg
 	return c
 }
 
+// Reset replaces the underlying iterator with a given one
+// and advanced it until the last known record ID is reached.
+//
+// It should be used only with tailable cursors.
+func (c *Cursor) Reset(iter types.DocumentsIterator) error {
+	if c.Type != Tailable && c.Type != TailableAwait {
+		panic("Reset called on non-tailable cursor")
+	}
+
+	c.m.Lock()
+
+	c.l.Debug("Resetting cursor")
+	c.iter = iter
+	recordID := c.lastRecordID
+
+	c.m.Unlock()
+
+	for {
+		_, doc, err := c.Next()
+		if err != nil {
+			return lazyerrors.Error(err)
+		}
+
+		if doc.RecordID() == recordID {
+			return nil
+		}
+	}
+}
+
 // Next implements types.DocumentsIterator interface.
 func (c *Cursor) Next() (struct{}, *types.Document, error) {
+	c.m.Lock()
+	defer c.m.Unlock()
+
+	if c.iter == nil {
+		return struct{}{}, nil, iterator.ErrIteratorDone
+	}
+
 	zero, doc, err := c.iter.Next()
 	if doc != nil {
+		recordID := doc.RecordID()
+		c.lastRecordID = recordID
+
 		if c.ShowRecordID {
-			doc.Set("$recordId", doc.RecordID())
+			doc.Set("$recordId", recordID)
 		}
 	}
 
 	return zero, doc, err
 }
 
 // Close implements types.DocumentsIterator interface.
+//
+// It closes the underlying iterator.
+// For normal cursors, it also removes it from the registry.
 func (c *Cursor) Close() {
-	c.closeOnce.Do(func() {
-		c.iter.Close()
-		c.iter = nil
+	c.m.Lock()
 
-		c.r.delete(c)
+	if c.iter == nil {
+		c.m.Unlock()
+		return
+	}
 
-		close(c.closed)
+	c.l.Debug("Closing cursor's iterator")
+	c.iter.Close()
+	c.iter = nil
+
+	c.m.Unlock()
+
+	// It is not entirely clear if we should do that; more tests are needed.
```

Comment:
@noisersup, do we have tests checking that normal cursors disappear after they are exhausted?

Reply:
Yep `TestCursorGetMoreCommand`/`GetMoreCursorExhausted`.